hwilib/devices/tapsigner.py

```diff
@@ -0,0 +1,408 @@
+"""
+Tapsigner
+********
+"""
+import os
+import pyaes
+import struct
+import getpass
+import datetime
+from base64 import b64encode
+from typing import List, Dict, Any, Union
+from cktap.transport import find_cards, CKTapUnixTransport, CKTapCard, CardRuntimeError  # type: ignore
+from cktap.utils import ser_compact_size, pick_nonce, path2str  # type: ignore
+from ..common import Chain, AddressType, hash256, hash160
+from ..key import ExtendedKey, get_bip44_purpose, get_bip44_chain
+from ..psbt import PSBT
+from ..descriptor import MultisigDescriptor
+from ..errors import UnavailableActionError, BadArgumentError, NO_PASSWORD, NoPasswordError
+from ..hwwclient import HardwareWalletClient
+from .._serialize import ser_sig_der, ser_string
+from .._script import (
+    is_p2pk,
+    is_p2pkh,
+    is_p2sh,
+    is_p2wpkh,
+    is_p2wsh,
+    is_witness,
+)
+
+
+def is_cvc_length_range(cvc: str) -> bool:
+    """CVC has to be at least 6 characters long and at most 32 characters long"""
+    return 6 <= len(cvc) <= 32
+
+
+def find_card_by_path(path: str) -> CKTapCard:
+    # path is actually an ident
+    for card in find_cards():
+        if path in card.card_ident:
+            return card
+    raise RuntimeError("Card not found! Is it on the reader?")
+
+
+class TapsignerClient(HardwareWalletClient):
+    # In path variable we store card ident
+    # In password variable we store card CVC
+    def __init__(self, path: str, password: str = "", expert: bool = False, chain: Chain = Chain.MAIN) -> None:
+        super(TapsignerClient, self).__init__(path, password, expert, chain)
+        self.device = find_card_by_path(path=path)
+        if not password:
+            raise NoPasswordError('Password (cvc) must be supplied for Tapsigner')
+
+    def get_master_xpub(self, addrtype: AddressType = AddressType.WIT, account: int = 0) -> ExtendedKey:
+        bip44_pupose = get_bip44_purpose(addrtype)
+        bip44_chain = get_bip44_chain(self.chain)
+        path = f"m/{bip44_pupose}h/{bip44_chain}h/{account}h"
+        return self.get_pubkey_at_path(path)
+
+    def get_master_fingerprint(self) -> bytes:
+        """
+        m fingerprint
+        """
+        extended_key_m = self.get_pubkey_at_path("m")
+        fingerprint = hash160(extended_key_m.pubkey)[:4]
+        return fingerprint
+
+    def get_pubkey_at_path(self, bip32_path: str) -> ExtendedKey:
+        """
+        Get the public key at the BIP 32 derivation path.
+
+        :param bip32_path: The BIP 32 derivation path
+        :return: The extended public key
+        """
+        xpub = self.device.derive_xpub_at_path(self.password, fullpath=bip32_path)
+        return ExtendedKey.deserialize(xpub)
+
+    def sign_tx(self, tx: PSBT) -> PSBT:
+        # completely stolen and adjusted from digitalbox (blind signing)
```

Comment:
Would it be possible for you to refactor this so that the sighashing code is shared rather than copy-pasting.

Reply:
added `get_sighash_tuples` as a member of `PSBT` class